### PR TITLE
feat(config-error): raise a warning issue for misconfigured `matchConfidence`

### PR DESCRIPTION
**Pull request by Gabriel-Ladzaretti**
## Changes
Raise a [new warning issue](https://github.com/ladzaretti/grouping-mc-testing/issues/49) whenever the `matchConfidence` `PackageRules` matcher is used without credentials.

Tested both for onboarding and onboarded repos.
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/package-rules/index.spec.ts [hunk 1]
```diff
@@ -626,58 +629,95 @@ describe('util/package-rules/index', () => {
...
+      const dep = {
+        depType: 'dependencies',
+        depName: 'a',
+        mergeConfidenceLevel: undefined,
+      };
+      const res = applyPackageRules({ ...config, ...dep });
+      expect(res.x).toBeUndefined();
+    });
+
+    it('throws when unauthenticated', () => {
```
Comment (Gabriel-Ladzaretti) [inline lib/util/package-rules/index.spec.ts:697]: pr diff-wise, this is the only real addition in this file. the weirdish diff is due to a newly added `describe` block for all other `matchConfidence` tests.

File: lib/constants/error-messages.ts [hunk 2]
```diff
@@ -62,3 +62,6 @@ export const INVALID_PATH = 'invalid-path';
 
 // PAGE NOT FOUND
 export const PAGE_NOT_FOUND_ERROR = 'page-not-found';
+
+// Missing api required credentials
```
Comment (HonkingGoose) [inline lib/constants/error-messages.ts:66]:
```suggestion
// Missing API required credentials
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/util/merge-confidence/index.ts [hunk 3]
```diff
@@ -191,6 +191,12 @@ export async function initMergeConfidence(): Promise<void> {
   return;
 }
 
+export function getApiToken(): string | undefined {
```
Comment (viceice) [inline lib/util/merge-confidence/index.ts:194]: why this code move?

File: lib/util/package-rules/merge-confidence.ts [hunk 4]
```diff
@@ -10,6 +12,18 @@ export class MergeConfidenceMatcher extends Matcher {
...
     }
+
+    /*
+     * Throw an error for unauthenticated use of the matchConfidence matcher.
+     */
+    if (is.undefined(getApiToken())) {
+      const error = new Error(MISSING_API_CREDENTIALS);
+      error.validationError = 'Missing credentials';
+      error.validationMessage =
+        "The 'matchConfidence' matcher in 'packageRules' requires authentication. Please refer to the documentation and add the required host rule.";
```
Comment (viceice) [inline lib/util/package-rules/merge-confidence.ts:23]: can we add a link to the docs?

File: lib/workers/repository/finalize/index.ts [hunk 5]
```diff
@@ -39,3 +37,11 @@ export async function finalizeRepo(
   runBranchSummary(config);
   runRenovateRepoStats(config, prList);
 }
+
+// istanbul ignore next
```
Comment (viceice) [inline lib/workers/repository/finalize/index.ts:41]: why ignore?

File: lib/util/merge-confidence/index.ts (hunk 3 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/merge-confidence/index.ts:194]: i prefer to have exported function at the top, but fact being that `queryApi` is above this and is un-exported, ill just move it back 😅

File: lib/workers/repository/finalize/index.ts (hunk 5 above)
Comment (Gabriel-Ladzaretti) [reply lib/workers/repository/finalize/index.ts:41]: the calling `finalizeRepo` function is ignored

https://github.com/renovatebot/renovate/blob/0ec88ed4352aaaa673969bb34ebf0e94edc97108/lib/workers/repository/finalize/index.ts#L14-L15

File: lib/util/package-rules/merge-confidence.ts (hunk 4 above)
Comment (Gabriel-Ladzaretti) [reply lib/util/package-rules/merge-confidence.ts:23]: We can do it "indirectly", but the result is not pretty.

For example
_____
There are missing credentials for the authentication-required feature. As a precaution, Renovate will pause PRs until it is resolved.

Error type: Missing credentials
Message: `The 'matchConfidence' matcher in 'packageRules' requires authentication. Please refer to the `[`documentation`](https://docs.renovatebot.com/configuration-options/#matchconfidence)` and add the required host rule.`
____

source markdown:
```markdown
 Please refer to the `[`documentation`](https://docs.renovatebot.com/configuration-options/#matchconfidence)` and add the required host rule.`
```

Comment (Gabriel-Ladzaretti) [reply lib/util/package-rules/merge-confidence.ts:23]: how about this style instead or something similar ?

____
There are missing credentials for the authentication-required feature. As a precaution, Renovate will pause PRs until it is resolved.

Error type: The `matchConfidence` matcher in `packageRules` requires authentication. Please refer to the [documentation](https://docs.renovatebot.com/configuration-options/#matchconfidence) and add the required host rule.
Message:  `Missing credentials`

___

Comment (viceice) [reply lib/util/package-rules/merge-confidence.ts:23]: @rarkins WDYT?

Comment (Gabriel-Ladzaretti) [reply lib/util/package-rules/merge-confidence.ts]: [Current issue iteration](https://github.com/ladzaretti/grouping-mc-testing/issues/49)

**Review by viceice — APPROVED** [review]

**Review by HonkingGoose — APPROVED** [review]

**Pull request merged**